cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1200,70 +1201,71 @@ template <>
+    if (!multiple || !multiple->is_valid) {
       return Status::Invalid("Rounding multiple must be non-null and valid");
     }
-    if (is_floating(type.id())) {
-      switch (options.multiple->type->id()) {
-        case Type::FLOAT: {
-          if (UnboxScalar<FloatType>::Unbox(*options.multiple) < 0) {
-            return Status::Invalid("Rounding multiple must be positive");
-          }
-          break;
-        }
-        case Type::DOUBLE: {
-          if (UnboxScalar<DoubleType>::Unbox(*options.multiple) < 0) {
-            return Status::Invalid("Rounding multiple must be positive");
-          }
-          break;
-        }
-        case Type::HALF_FLOAT:
-          return Status::NotImplemented("Half-float values are not supported");
-        default:
-          return Status::Invalid("Rounding multiple must be a ", type, " scalar, not ",
-                                 *options.multiple->type);
-      }
+
+    // Ensure the rounding multiple option matches the kernel's output type.
+    // The output type is not available here so we use the following rule:
+    // If `multiple` is neither a floating-point nor a decimal type, then
+    // cast to float64, else cast to the kernel's input type.
+    std::shared_ptr<Scalar> resolved_multiple;
+    const auto& to_type =
+        (!is_floating(multiple->type->id()) && !is_decimal(multiple->type->id()))
+            ? float64()
+            : args.inputs[0].type;
+    bool is_casted = false;
+    if (!multiple->type->Equals(to_type)) {
+      ARROW_ASSIGN_OR_RAISE(
+          auto casted_multiple,
+          Cast(Datum(multiple), to_type, CastOptions::Safe(), ctx->exec_context()));
+      resolved_multiple = casted_multiple.scalar();
+      is_casted = true;
     } else {
-      DCHECK(is_decimal(type.id()));
-      if (!type.Equals(*options.multiple->type)) {
-        return Status::Invalid("Rounding multiple must be a ", type, " scalar, not ",
-                               *options.multiple->type);
-      }
-      switch (options.multiple->type->id()) {
-        case Type::DECIMAL128: {
-          if (UnboxScalar<Decimal128Type>::Unbox(*options.multiple) <= 0) {
-            return Status::Invalid("Rounding multiple must be positive");
-          }
-          break;
-        }
-        case Type::DECIMAL256: {
-          if (UnboxScalar<Decimal256Type>::Unbox(*options.multiple) <= 0) {
-            return Status::Invalid("Rounding multiple must be positive");
-          }
-          break;
-        }
-        default:
-          // This shouldn't happen
-          return Status::Invalid("Rounding multiple must be a ", type, " scalar, not ",
-                                 *options.multiple->type);
-      }
+      resolved_multiple = multiple;
     }
-    return std::move(state);
+
+    // NOTE: The positive value check can be simplified by using a comparison kernel.
+    bool is_negative = false;
+    switch (resolved_multiple->type->id()) {
```

Comment:
An interesting way of writing this would be to use `VisitScalarInline`:
```c++
bool IsNegative(const Scalar& scalar) {
  struct IsNegativeVisitor {
    bool result = false;

    template <typename... Ts>
    Status Visit(const NumericScalar<Ts...>& scalar) {
      result = scalar.value < 0;
      return Status::OK();
    }
    template <typename... Ts>
    Status Visit(const DecimalScalar<Ts...>& scalar) {
      result = scalar.value < 0;
      return Status::OK();
    }
    Status Visit(const Scalar& scalar) {
      return Status::OK();
    }
  };
  IsNegativeVisitor visitor{};
  std::ignore = VisitScalarInline(scalar, &visitor);
  return visitor.result;
}
```

Reply:
Yeah, I thought about something like this but couldn't wrap my head around on where to limit its functionality. For example, `IsNegative` could support all Arrow types so that it can be used as an idiom for such comparisons. But then, we can also have `IsPositive`, `IsZero`, or better yet get the threshold value as a parameter. But then it begins to look like the comparison kernels. So why not use the comparison kernels via `Op::Call()`? At the moment these are not visible.

Reply:
I think it would be great if primitive kernels (arithmetic, logical, comparisons) were organized in such a way that they can be reused throughout the codebase.

Reply:
BTW, I do like the visitor pattern and `xxx.value` idiom.

Reply:
I think we want to avoid calling a kernel for a simple operation on a scalar, as the kernel execution overhead is large.

Reply:
Ok, I am simply trying to reuse code when possible.
I added a `IsPositive` visitor, but had to move the struct outside of the function (`templates are not allowed in local classes`).